File: paper/src/main/java/com/badbones69/crazycrates/support/structures/StructureHandler.java
package com.badbones69.crazycrates.support.structures;

import com.badbones69.crazycrates.CrazyCrates;
import com.google.common.collect.Lists;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.structure.Mirror;
import org.bukkit.block.structure.StructureRotation;
import org.bukkit.structure.StructureManager;
import org.bukkit.util.BlockVector;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class StructureHandler {

    private final File file;

    private final CrazyCrates plugin = CrazyCrates.getPlugin();

    public StructureHandler(File file) {
        this.file = file;
    }

    private final List<Location> structureBlocks = new ArrayList<>();

    private final List<Location> preStructureBlocks = new ArrayList<>();

    private StructureManager getStructureManager() {
        return plugin.getServer().getStructureManager();
    }

    private BlockVector getStructureSize() {
        try {
            return getStructureManager().loadStructure(file).getSize();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void pasteStructure(Location location) {
        try {
            getNearbyBlocks(location);

            getStructureManager().loadStructure(file).place(location, false, StructureRotation.NONE, Mirror.NONE, 0, 1F, new Random());

            getBlocks(true, location);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void removeStructure() {
        structureBlocks.forEach(block -> {
            Location blockLoc = block.toBlockLocation();

            blockLoc.getBlock().setType(Material.AIR, true);
        });
    }

    public void saveStructure(Location location) {

    }

    public double getStructureX() {
        try {
            return getStructureSize().getX();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return 0;
    }

    public double getStructureZ() {
        try {
            return getStructureSize().getZ();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return 0;
    }

    private void getBlocks(boolean getStructureBlocks, Location location) {
        for (int x = 0; x < getStructureSize().getX(); x++) {
            for (int y = 0; y < getStructureSize().getY(); y++) {
                for (int z = 0; z < getStructureSize().getZ(); z++) {

                    Block relativeLocation = location.getBlock().getRelative(x, y, z);

                    List<Location> relativeBlocks = new ArrayList<>();

                    relativeBlocks.add(relativeLocation.getLocation());

                    if (getStructureBlocks) {
                        structureBlocks.addAll(relativeBlocks);

                        structureBlocks.forEach(block -> {
                            Location blockLoc = block.toBlockLocation();

                            blockLoc.getBlock().getState().update();
                        });

                        continue;
                    }

                    preStructureBlocks.addAll(relativeBlocks);
                }
            }
        }
    }

    public List<Location> getNearbyBlocks(Location location) {
        getBlocks(false, location);

        return preStructureBlocks;
    }

    public List<Material> getBlockBlackList() {
        return Lists.newArrayList(Material.ACACIA_SIGN, Material.BIRCH_SIGN, Material.DARK_OAK_SIGN, Material.JUNGLE_SIGN, Material.OAK_SIGN,
                Material.SPRUCE_SIGN, Material.ACACIA_WALL_SIGN, Material.BIRCH_WALL_SIGN, Material.DARK_OAK_WALL_SIGN, Material.JUNGLE_WALL_SIGN, Material.OAK_WALL_SIGN,
                Material.SPRUCE_WALL_SIGN,Material.STONE_BUTTON,Material.BIRCH_BUTTON,Material.ACACIA_BUTTON,Material.DARK_OAK_BUTTON, Material.JUNGLE_BUTTON, Material.SPRUCE_BUTTON);
    }
}

File: plugin/src/main/java/com/badbones69/crazycrates/v2/utils/quadcrates/StructureHandler.kt
package com.badbones69.crazycrates.v2.utils.quadcrates

import com.badbones69.crazycrates.getPlugin
import com.badbones69.crazycrates.v2.utils.interfaces.Structures
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.structure.Mirror
import org.bukkit.block.structure.StructureRotation
import org.bukkit.entity.Player
import java.io.File
import java.util.*
import kotlin.collections.ArrayList

class StructureHandler(val file: File) : Structures {

    private val structureManager = getPlugin().server.structureManager.loadStructure(file)

    private val blockLocation = ArrayList<Block>()

    private val blocks = ArrayList<Block>()

    // Structure pastes from the corner on the player.
    // Adding -2.2, 0.0 -2.2 to the "location" somewhat centers it but varies from where the player is standing.
    override fun pasteStructure(location: Location) {
        runCatching {
            //.add(-2.2, 0.0, -2.2)
            structureManager.place(location, false, StructureRotation.NONE, Mirror.NONE, 0, 1F, Random())
        }.onFailure { getPlugin().logger.warning(it.message) }
    }

    // Don't need
    override fun saveStructure(location: ArrayList<Location>) {
        runCatching {

        }.onFailure { getPlugin().logger.warning(it.message) }
    }

    // Got to paste it properly first lol.
    override fun removeStructure(location: Location) {
        runCatching {
            blockLocation.forEach { println(it.type) }
        }.onFailure { getPlugin().logger.warning(it.message) }
    }

    override fun getStructureBlocks(location: Location): ArrayList<Block> {
        for (x in 0 until structureManager.size.x.toInt()) {
            for (y in 0 until structureManager.size.y.toInt()) {
                for (z in 0 until structureManager.size.z.toInt()) {
                    blockLocation.add(location.block.getRelative(x, y, z))
                }
            }
        }
        return blockLocation
    }

    override fun getNearbyBlocks(location: Location): ArrayList<Block> {
        for (x in 0 until structureManager.size.x.toInt()) {
            for (y in 0 until structureManager.size.y.toInt()) {
                for (z in 0 until structureManager.size.z.toInt()) {
                    blocks.add(location.block.getRelative(x, y, z))
                }
            }
        }
        return blocks
    }

    override fun getBlackList(): List<Material> {
        return arrayListOf(
            Material.ACACIA_BUTTON,
            Material.BIRCH_BUTTON,
            Material.DARK_OAK_BUTTON,
            Material.JUNGLE_BUTTON,
            Material.OAK_BUTTON,
            Material.SPRUCE_BUTTON,
            Material.STONE_BUTTON,
            // Add all signs
            Material.ACACIA_SIGN,
            Material.BIRCH_SIGN,
            Material.DARK_OAK_SIGN,
            Material.JUNGLE_SIGN,
            Material.OAK_SIGN,
            Material.SPRUCE_SIGN,
            Material.ACACIA_WALL_SIGN,
            Material.BIRCH_WALL_SIGN,
            Material.DARK_OAK_WALL_SIGN,
            Material.JUNGLE_WALL_SIGN,
            Material.OAK_WALL_SIGN,
            Material.SPRUCE_WALL_SIGN
        )
    }
}